ThirdPartyAdapters/chartboost/chartboost/src/main/java/com/google/ads/mediation/chartboost/ChartboostAdapter.java

```diff
@@ -186,6 +186,101 @@ public void onPause() {
     public void onResume() {
     }
 
+    @Override
+    public void requestBannerAd(Context context,
+                                MediationBannerListener mediationBannerListener,
+                                Bundle serverParameters,
+                                AdSize adSize,
+                                MediationAdRequest mediationAdRequest,
+                                Bundle networkExtras) {
+        mMediationBannerListener = mediationBannerListener;
+        mChartboostParams = ChartboostAdapterUtils.createChartboostParams(serverParameters, networkExtras);
+        if (!ChartboostAdapterUtils.isValidChartboostParams(mChartboostParams)) {
+            // Invalid server parameters, send ad failed to load event.
+            if (mMediationBannerListener != null) {
+                mMediationBannerListener.onAdFailedToLoad(ChartboostAdapter.this,
+                        AdRequest.ERROR_CODE_INVALID_REQUEST);
+            }
+            return;
+        }
+        ChartboostSingleton.startChartboostBanner(context, mChartboostBannerDelegate, createChartboostBannerListener());
```

Comment:
I think this would be a good moment to move away from the singleton pattern, since banners are already modular. Implementation would be a lot simpler. But I'm fine with following the existing pattern and doing this later with modularization if you prefer that.

Reply:
First I wanted to follow current patter to understand what is going on, but Im not very happy with this implementation. Im not sure also i've got time to break it apart and change the way it is done. I could spend few more days on it. Let's see